File: chaincode-go/updatable_encryption_go/key_generation.go
package main

import "fmt"

func key_gen(R1 [][]int) ([][]int, [][]int, [][]int, [][]int, error) {
	A0 := sample_uniform_matrix(C.n, C.m_tilda)
	var R1_new [][]int
	if R1 == nil {
		R1_new = sample_normal_matrix(C.m_tilda, C.nk, C.mean, C.sigma_sk)
		VerifyR(R1_new)
	} else {
		R1_new = R1
	}

	R2 := sample_normal_matrix(C.m_tilda, C.nk, C.mean, C.sigma_sk)
	VerifyR(R2)

	A1, err1 := dot_product_MM(A0, R1_new)

	if err1 != nil {
		return nil, nil, nil, nil, fmt.Errorf("key_gen - err1 %v", err1)
	}

	A1 = negate_matrix_values(A1)

	A2, err2 := dot_product_MM(A0, R2)

	if err2 != nil {
		return nil, nil, nil, nil, fmt.Errorf("key_gen - err2 %v", err2)
	}

	A2 = negate_matrix_values(A2)

	/*
			# to be removed
		    # verify trapdoor
		    R = np.block([R1, R2])
		    pk_local = np.block([A0, A1, A2])
		    R_ext = np.block([[R], [np.eye(R.shape[1], dtype=int)] ])
		    Res = np.dot(pk_local, R_ext) % q
		    all_zeros = not np.any(Res)
		    assert all_zeros
	*/

	return A0, A1, A2, R1_new, nil
}

File: src/updatable_encryption_python/key_generation.py
import numpy as np
from updatable_encryption_python.parameters import *
from updatable_encryption_python.utility_functions import *

def KeyGen(sk1):
    # uniform distribution with lower and upper bound q
    A0 = sample_uniform_matrix(n, m_tilda, q)

    # normal distribution
    if sk1 is None:
        R1 = sample_normal_matrix(m_tilda, nk, mean, sigma_sk)
        VerifyR(R1)
    else:
        R1 = sk1
    R2 = sample_normal_matrix(m_tilda, nk, mean, sigma_sk)
    VerifyR(R2)
    
    A1 = -np.dot(A0, R1) % q
    A2 = -np.dot(A0, R2) % q
    
    pk = [A0, A1, A2]  # Public Key
    #pk_new = np.block([A0, A1, A2])
    sk = R1

    # to be removed
    # verify trapdoor
    R = np.block([R1, R2])
    pk_local = np.block([A0, A1, A2])
    R_ext = np.block([[R], [np.eye(R.shape[1], dtype=int)] ])
    Res = np.dot(pk_local, R_ext) % q
    all_zeros = not np.any(Res)
    assert all_zeros

    #print("KeyGen sk=", sk)
    return pk, sk
